incorrect recusrive struct detection

The following program `sample.go` triggers a panic:

```go
package main

type S struct {
	t *T
}

func newS() *S {
	return &S{
		t: &T{u: map[string]*U{}},
	}
}

type T struct {
	u map[string]*U
}

type U struct {
	a int
}

func main() {
	s := newS()
	_ = s

	println("ok")
}

```

Expected result:
```console
$ go run ./sample.go
ok
```

Got:
```console
$ yaegi ./sample.go
_test/test.go:9:12: panic
_test/test.go:22:7: panic
reflect.Set: value of type *struct { Xu map[string]*struct { Xa int } } is not assignable to type *struct { Xu map[string]*interface {} }
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0002e4600, 0x40)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc000155de0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/interp.go:333 +0xbc
panic(0x191b9e0, 0xc00018ac40)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc0003640a0, 0xc000372200)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:116 +0x20c
panic(0x191b9e0, 0xc00018ac40)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc000364140, 0xc000370200)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:116 +0x20c
panic(0x191b9e0, 0xc00018ac40)
	/usr/local/Cellar/go/1.14.4/libexec/src/runtime/panic.go:969 +0x166
reflect.Value.assignTo(0xc000184d00, 0xc00018c330, 0x196, 0x1a57b55, 0xb, 0xc000184c00, 0x0, 0x1014174, 0xc000155980, 0x109ee93)
	/usr/local/Cellar/go/1.14.4/libexec/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0xc000184c00, 0xc00018c2b0, 0x196, 0xc000184d00, 0xc00018c330, 0x196)
	/usr/local/Cellar/go/1.14.4/libexec/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.doCompositeSparse.func1(0xc000364140, 0xc000192370)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:2112 +0x160
github.com/containous/yaegi/interp.runCfg(0xc000370200, 0xc000364140)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:122 +0x6d
github.com/containous/yaegi/interp.call.func7(0xc0003640a0, 0x10b443d)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:1008 +0xc42
github.com/containous/yaegi/interp.runCfg(0xc000372200, 0xc0003640a0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:122 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0001b21a0, 0xc000371a00, 0xc000364000)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/run.go:100 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0001b21a0, 0xc00036a000, 0xda, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/interp/interp.go:409 +0x52b
main.main()
	/Users/nrwiersma/Go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:159 +0x65d
```

--- a/_test/struct54.go
+++ b/_test/struct54.go
@@ -1,0 +1,26 @@
+package main
+
+type S struct {
+	t *T
+}
+
+func newS() *S {
+	return &S{
+		t: &T{u: map[string]*U{}},
+	}
+}
+
+type T struct {
+	u map[string]*U
+}
+
+type U struct {
+	a int
+}
+
+func main() {
+	s := newS()
+	_ = s
+
+	println("ok")
+}

--- a/interp/type.go
+++ b/interp/type.go
@@ -1260,7 +1260,7 @@ func (t *itype) refType(defined map[string]*itype, wrapRecursive bool) reflect.T
 	if defined[name] != nil && defined[name].rtype != nil {
 		return defined[name].rtype
 	}
-	if t.val != nil && t.val.cat == structT && t.val.rtype == nil && hasRecursiveStruct(t.val, defined) {
+	if t.val != nil && t.val.cat == structT && t.val.rtype == nil && hasRecursiveStruct(t.val, copyDefined(defined)) {
 		// Replace reference to self (direct or indirect) by an interface{} to handle
 		// recursive types with reflect.
 		typ := *t.val
@@ -1365,6 +1365,14 @@ func (t *itype) implements(it *itype) bool {
 		return t.TypeOf().Implements(it.TypeOf())
 	}
 	return t.methods().contains(it.methods())
+}
+
+func copyDefined(m map[string]*itype) map[string]*itype {
+	n := make(map[string]*itype, len(m))
+	for k, v := range m {
+		n[k] = v
+	}
+	return n
 }
 
 // hasRecursiveStruct determines if a struct is a recursion or a recursion

